### PR TITLE
[CPDTP-441] - Add defer API endpoint to NPQ

### PR DESCRIPTION
**Pull request by StupidCodeFactory**
## Ticket and context

Ticket: https://dfedigital.atlassian.net/browse/CPDTP-441

### Is there anything that the code reviewer should know?
I've continued going down the road extrapolating deferral like I did before for withdrawal. 
I will keep following the same pattern for resuming an application.

### Code quality checks
- [x] All commit messages are meaningful and true
- [x] Added enough automated tests

### HTML Checks
- [ ] All new pages have automated accessibility checks
- [ ] All new pages have visual tests via Percy

### CODE REVIEW

File: lib/api/participant_actions/routing.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  module ParticipantActions
+    class Routing
+      def call(mapper, _options = {})
+        mapper.member { mapper.put :withdraw }
+        mapper.member { mapper.put :defer }
+      end
+
+    private
+
+      attr_accessor :defaults
```
Comment (kicferk1) [inline lib/api/participant_actions/routing.rb:13]: What's this attribute for?

**Review by kicferk1 — COMMENTED** [review]
We need some api documentation for it, right? :)

File: app/models/participant_profile_state.rb [hunk 2]
```diff
@@ -27,4 +13,30 @@ def valid?(_context = nil)
   }
 
   scope :most_recent, -> { order("created_at desc").limit(1) }
+
+private
+
+  def withdrawal
+    return unless current_state
+
+    errors.add(:base, I18n.t(:invalid_withdrawal)) if current_state.state == self.class.states[:withdrawn]
```
Comment (kicferk1) [inline app/models/participant_profile_state.rb:22]: `errors.add(:base, I18n.t(:invalid_withdrawal)) if current_state.withdrawn?`?

Likewise for other conditionals in this file :)

File: app/services/participants/defer/validate_and_change_state.rb [hunk 3]
```diff
@@ -6,11 +6,19 @@ module ValidateAndChangeState
       extend ActiveSupport::Concern
       include ActiveModel::Validations
 
+      REASONS = %w[
+        bereavement
+        long-term-sickness
+        parental-leave
+        career-break
+        other
+      ].freeze
```
Comment (kicferk1) [inline app/services/participants/defer/validate_and_change_state.rb:15]: Are the deferral reasons the same for ecf and npq?

File: app/services/participants/withdraw/validate_and_change_state.rb [hunk 4]
```diff
@@ -5,11 +5,18 @@ module Withdraw
     module ValidateAndChangeState
       extend ActiveSupport::Concern
       include ActiveModel::Validations
+      REASONS = %w[
+        left-teaching-profession
+        moved-school
+        mentor-no-longer-being-mentor
+        school-left-fip
+        other
+      ].freeze
```
Comment (kicferk1) [inline app/services/participants/withdraw/validate_and_change_state.rb:14]: Withdraw reasons certainly are not the same between ecf and npq. Does this change not cause NPQ tests for withdrawals to fail? Or does the constant included in the npq module override this constant?

File: spec/docs/npq_participants_spec.rb [hunk 5]
```diff
@@ -56,6 +56,16 @@
...
 
+  it_behaves_like "JSON Participant Deferral documentation", "/api/v1/participants/npq/{id}/defer" do
+    let(:participant) { npq_application }
+    let(:attributes) do
+      {
+        reason: Participants::Defer::NPQ::REASONS.sample,
+        course_identifier: npq_application.npq_course.identifier,
+      }
+    end
+  end
```
Comment (kicferk1) [inline spec/docs/npq_participants_spec.rb:67]: I don't think the npq endpoint should have the same schemas as the ecf ones - which it currently does with this shared examples setup.

File: app/services/participants/defer/validate_and_change_state.rb (hunk 3 above)
Comment (StupidCodeFactory) [reply app/services/participants/defer/validate_and_change_state.rb:15]: yes

**Comment by StupidCodeFactory** [comment]
> Let's add something to release notes and usage guide about this endpoint please

I'd happily do that but I've got no clue where release notes are, would you be able to point me in the direction?

**Pull request merged**